### DISCUSSION TITLE
Problems on self-hosted runner .. idea?

### DISCUSSION THREAD

**Discussion by robertklaus-bosch** [whelk-io/maven-settings-xml-action#97] [General]
I am trying this action (v14, v16) on _GitHub Enterprise_ (version 2.x) with a _self-hosted runner_ (not under my control).

Problem is, that I never get this to work as the runner setup is different from the GitHub cloud runners.
When using ```stCarolas/setup-maven@v4``` the settings are never found as the maven installation does not pick up the settings from the user home.
When I try to configre the settings with ```--settings``` I get a stacktrace as below.

Only thing I can think of is that maybe it helps if this action could (optionally) generate the settings.xml in a folder specified in the workflow. Maybe that would work.

Any ideas how to solve this?

```
[DEBUG] Message scheme: color
[DEBUG] Message styles: debug info warning error success failure strong mojo project
[DEBUG] Reading global settings from /home/<user>/actions-runner/_work/_tool/maven/3.8.1/x64/conf/settings.xml
[DEBUG] Reading user settings from /home/<user>/.m2/settings.xml
[DEBUG] Reading global toolchains from /home/<user>/actions-runner/_work/_tool/maven/3.8.1/x64/conf/toolchains.xml
[DEBUG] Reading user toolchains from /home/<user>/.m2/toolchains.xml
[DEBUG] Using local repository at /home/<user>/.m2/repository
[DEBUG] Using manager EnhancedLocalRepositoryManager with priority 10.0 for /home/<user>/.m2/repository
[DEBUG] Failed to decrypt password for server github: org.sonatype.plexus.components.sec.dispatcher.SecDispatcherException: java.io.FileNotFoundException: /home/<user>/.m2/settings-security.xml (No such file or directory)
```
(Note replaced actual user with ```<user>```)

**Comment by zteater** [comment]
@robertklaus-bosch sounds like you're using an encrypted password in your `settings.xml`, which requires `settings-security.xml` to decrypt. This action unfortunately only creates the `settings.xml`.  [See maven encryption for more information](https://maven.apache.org/guides/mini/guide-encryption.html)

However, I see 2 paths forward for you:
1) we could create a new action to generate a `settings-security.xml`, in a similar manner to this action, or
2) use GH secrets to inject a plain-text password into `settings.xml`

```yaml
- name: maven-settings-xml-action
  uses: whelk-io/maven-settings-xml-action@v16
  with:
    repositories: |
      [
        { 
          "id": "some-repository", 
          "url": "http://some.repository.url" 
        }
      ]
    servers: |
      [
         { 
           "id": "github", 
           "username": ${{ secrets.MY_USERNAME }}, 
           "password": ${{ secrets.MY_PASSWORD }} 
         }
      ]
```

**Reply by robertklaus-bosch** [reply]
@zteater actually I'm trying to avoid having passwords in the generated file and use environment variables instead, which will then contain the secrets from GitHub. At least that is the idea.
Any idea why Maven then thinks this contains encrypted passwords?
Snippet looks like this:

```
name: Create settings.xml
uses: whelk-io/maven-settings-xml-action@v16
with:
	repositories: '[{ "id": "<someId>", "url": "https://<somehost>/<somerepo>" }]'
	plugin_repositories: '[{ "id": "<someId>", "url": "https://<somehost>/<somerepo>" }]'
	servers: '[{ "id": "<someId>", "username": "${env.DEPLOY_USER_REF}", "password": "${env.DEPLOY_TOKEN_REF}" }]'
	mirrors: '[{ "id": "Artifactory-Central-Mirror", "mirrorOf": "external:*", "url": "https://<somehost>/<somerepo>" }]'
```

**Reply by zteater** [reply]
@robertklaus-bosch sounds like you're setting a `DEPLOY_TOKEN_REF` from a GH Secrets. Is the value in secrets encrypted or plain-text? The maven encrypted passwords look like like a base64 encoding with `{ }`, e.g. `{jSMOWnoPFgsHVpMvz5VrIt5kRbzGpI8u+9EF1iFQyJQ=}`. If your password is encapsulated with `{ }` I believe Maven would think it is an encrypted password and try to decrypt it with settings-security.xml.

**Reply by zteater** [reply]
@robertklaus-bosch added some unit tests to confirm env variables are populated correctly in the settings.xml: https://github.com/whelk-io/maven-settings-xml-action/pull/109

**Reply by zteater** [reply]
@robertklaus-bosch I'm guessing the env variable doesn't have a value. 

https://maven.apache.org/guides/mini/guide-encryption.html#how-to-encrypt-server-passwords

The maven settings allows for anything outside of the `{ }`, e.g.
```xml
    <server>
      <id>my.server</id>
      <username>foo</username>
      <password>Oleg reset this password on 2009-03-11, expires on 2009-04-11 {COQLCE6DU6GtcS5P=}</password>
    </server>
```
So if your env variable is not set, then the `${env.DEPLOY_TOKEN_REF}` might persist as a string literal and maven interprets this as an encrypted password because of the `{ }`.

**Reply by robertklaus-bosch** [reply]
@zteater many thanks, I guess this might be the case. The order of the steps is to first setup Java, Maven, settings.xml then provide secrets. I think I'll try to set the variables first, maybe that'll do the trick. Gonna need a moment for that.

**Reply by robertklaus-bosch** [reply]
@zteater I set the environment variables and re-did my test. As it again failed I just used ```iamateapot``` instead of ```${env.DEPLOY_USER_REF}``` and ```${env.DEPLOY_TOKEN_REF}``` and got the same exception.
I would think, there is no reason that Maven thinks it contains encrypted passwords.

This is weird. As I'm using an action to setup Maven, I don't think it is configured to use a master password as well.
I'll try to give it a settings.xml in the workspace and see what it does then.

**Reply by robertklaus-bosch** [reply]
I tested following:
- added a ```settings.xml``` file as part of the repository (all credentials are non-sense but don't contain ```{ }```)
- removed ```whelk-io/maven-settings-xml-action@v16```
- set Maven to use the ```settings.xml``` with ```--settings```

That now could read / use the settings.xml then failed because it could not authenticate (obviously).

So in the generated ```settings.xml``` from ```maven-settings-xml-action``` there is something unexpected leading to Maven believe it has to deal with encrypted passwords or it actually might have to do with the setup.
Some questions:
- I don't think the action allows to see the output anywhere, right?
- maybe it is possible that the action is not allowed to write to the user home. What would happen in this case?

**Reply by zteater** [reply]
> I don't think the action allows to see the output anywhere, right?

It should be writing to `{user-home}/.m2/settings.xml`. Try adding this step after the action to see the generated `settings.xml` in your action logs.

```yaml
      - run: cat ~/.m2/settings.xml
```

> maybe it is possible that the action is not allowed to write to the user home. What would happen in this case?

It should throw an uncaught exception and break the step.

**Reply by robertklaus-bosch** [reply]
I dumped the ```settings.xml``` generated by ```maven-settings-xml-action``` then copied the content to the test file in the repo.
I did some formatting for readability, otherwise it is the generated content including ```${env.DEPLOY_USER_REF}```.
When using that with ```--settings``` Maven again does not attempt to decrypt and it works!

I hope this is not a formatting issue.

Would it be possible to extend this action with the possibility to write the file somewhere else - keeping the default as it is for compatibility?
I write somewhere in the github workspace and then pass it to Maven.

**Reply by zteater** [reply]
You're suggesting a feature where you could define a file path for the output instead of the default `~/.m2/settings.xml`?

**Reply by robertklaus-bosch** [reply]
Yes I think so. I think extending the action with an optional ```outfile``` might do the trick.
How to go on from here? Should I create an issue?

**Reply by zteater** [reply]
Added #111 for this feature. Feel free to drop a PR or I can pick it up maybe later today or tomorrow.

**Reply by robertklaus-bosch** [reply]
Sorry, was a little out of touch. Thanks for the issue and the commit. I'll help with testing if possible.

**Reply by zteater** [reply]
@robertklaus-bosch what do you think about this:

```yaml
- uses: whelk-io/maven-settings-xml-action@v18
  with:
    output_file: foo/custom.xml
```

To generate the following file in the home directory: `/home/runner/foo/custom.xml`

Which can be referenced in your maven steps like this: `mvn -s /home/runner/foo/custom.xml test`

**Reply by robertklaus-bosch** [reply]
I like the ```output_file``` - I would just like to use ```GITHUB_WORKSPACE``` if possible, maybe like so:
````
- uses: whelk-io/maven-settings-xml-action@v18
  with:
    output_file:  ${{ env.GITHUB_WORKSPACE }}/foo/custom.xml
````
then using ```mvn -s ${{ env.GITHUB_WORKSPACE }}/foo/custom.xml```

Makes sense?

**Reply by zteater** [reply]
ok, so the GH env variables were a bit trickey, but I got it worked out. I had to make sure node was resolving the env variables and we have to pass them slightly different:

```yaml
    steps:
      - uses: actions/checkout@v2.3.4
      - uses: whelk-io/maven-settings-xml-action@v18
        with:
          output_file: $GITHUB_WORKSPACE/custom.xml
      - uses: actions/setup-java@v1.4.3
        with:
          java-version: 11
      - run: |
          ./mvnw -s $GITHUB_WORKSPACE/custom.xml clean test
```

**Reply by robertklaus-bosch** [reply]
Tested today with version/branch ```feature/custom-output-file-path``` and it worked 👍
(```$GITHUB_WORKSPACE/custom.xml``` and ```$GITHUB_WORKSPACE/foo/custom.xml```)

**Reply by zteater** [reply]
Went ahead and published v18. Thank you for the feedback!
